File: erlang/anagram/src/anagram.erl
-module(anagram).

-export([find_anagrams/2]).

find_anagrams(Subject, Candidates) ->
    Word = string:casefold(Subject),
    Letters = lists:sort(Word),
    lists:filter(fun(Candidate) -> string:casefold(Candidate) =/= Word andalso Letters == lists:sort(string:casefold(Candidate)) end, Candidates).

File: ruby/anagram/anagram.rb
class Anagram
  private

  attr_reader :rejections, :selections

  def initialize(word)
    base = word.downcase
    letters = base.chars.sort
    @rejections = ->(candidate) { candidate.downcase == base }
    @selections = ->(candidate) { letters == candidate.downcase.chars.sort }
  end

  public

  def match(candidates)
    candidates.reject(&rejections)
              .select(&selections)
  end
end
